pkg/logentry/stages/replace.go

```diff
@@ -0,0 +1,218 @@
+	expression *regexp.Regexp
+	logger     log.Logger
+}
+
+// newReplaceStage creates a newReplaceStage
+func newReplaceStage(logger log.Logger, config interface{}) (Stage, error) {
+	cfg, err := parseReplaceConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	expression, err := validateReplaceConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &replaceStage{
+		cfg:        cfg,
+		expression: expression,
+		logger:     log.With(logger, "component", "stage", "type", "replace"),
+	}, nil
+}
+
+// parseReplaceConfig processes an incoming configuration into a ReplaceConfig
+func parseReplaceConfig(config interface{}) (*ReplaceConfig, error) {
+	cfg := &ReplaceConfig{}
+	err := mapstructure.Decode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// Process implements Stage
+func (r *replaceStage) Process(labels model.LabelSet, extracted map[string]interface{}, t *time.Time, entry *string) {
+	// If a source key is provided, the replace stage should process it
+	// from the extracted map, otherwise should fallback to the entry
+	input := entry
+
+	if r.cfg.Source != nil {
+		if _, ok := extracted[*r.cfg.Source]; !ok {
+			if Debug {
+				level.Debug(r.logger).Log("msg", "source does not exist in the set of extracted values", "source", *r.cfg.Source)
+			}
+			return
+		}
+
+		value, err := getString(extracted[*r.cfg.Source])
+		if err != nil {
+			if Debug {
+				level.Debug(r.logger).Log("msg", "failed to convert source value to string", "source", *r.cfg.Source, "err", err, "type", reflect.TypeOf(extracted[*r.cfg.Source]))
+			}
+			return
+		}
+
+		input = &value
+	}
+
+	if input == nil {
+		if Debug {
+			level.Debug(r.logger).Log("msg", "cannot parse a nil entry")
+		}
+		return
+	}
+
+	// Get string of matched captured groups. We will use this to extract all named captured groups
+	match := r.expression.FindStringSubmatch(*input)
+
+	matchAllIndex := r.expression.FindAllStringSubmatchIndex(*input, -1)
+
+	if matchAllIndex == nil {
+		if Debug {
+			level.Debug(r.logger).Log("msg", "regex did not match", "input", *input, "regex", r.expression)
+		}
+		return
+	}
+
+	// All extracted values will be available for templating
+	td := r.getTemplateData(extracted)
+
+	// Initialize the template with the "replace" string defined by user
+	templ, err := template.New("pipeline_template").Funcs(functionMap).Parse(r.cfg.Replace)
+	if err != nil {
+		if Debug {
+			level.Debug(r.logger).Log("msg", "template initialization error", "err", err)
+		}
+		return
+	}
+
+	result, capturedMap, err := r.getReplacedEntry(matchAllIndex, *input, td, templ)
+	if err != nil {
+		if Debug {
+			level.Debug(r.logger).Log("msg", "failed to execute template on extracted value", "err", err)
+		}
+		return
+	}
+
+	if r.cfg.Source != nil {
+		extracted[*r.cfg.Source] = result
+	} else {
+		*entry = result
+	}
+
+	// All the named captured group will be extracted
+	for i, name := range r.expression.SubexpNames() {
```

Comment:
I'm not sure this is useful ?

Reply:
A lot of you code would be much simpler.

Let's drop this for now. You can have a regex to do that.

Reply:
> Let's drop this for now. You can have a regex to do that

This is just for reducing using different stages in pipeline. If user wants to extract variable and then template the extracted variable he should use regex stage and then template stage. This would reduce that effort

> A lot of you code would be much simpler.

Sure. Let me know where I have to improve.

Reply:
Let's start with this.